orderbook/src/trade_events.rs

```diff
@@ -0,0 +1,147 @@
+use crate::database::{Database, Trade as DbTrade};
+use anyhow::{anyhow, Context, Error, Result};
+use contracts::{
+    g_pv_2_settlement::{event_data::Trade as ContractTrade, Event as ContractEvent},
+    GPv2Settlement,
+};
+use ethcontract::{errors::ExecutionError, Event, EventMetadata};
+use futures::{Stream, StreamExt, TryStreamExt};
+use model::order::OrderUid;
+use std::{convert::TryInto, ops::RangeInclusive, sync::Arc};
+use web3::{Transport, Web3};
+
+// We expect that there is never a reorg that changes more than the last n blocks.
+const MAX_REORG_BLOCK_COUNT: u64 = 25;
+// When we insert new trade events into the database we will insert at most this many in one
+// transaction.
+const INSERT_TRADE_BATCH_SIZE: usize = 250;
+
+pub struct TradeEvents {
+    contract: GPv2Settlement,
+    db: Arc<dyn Database>,
```

Comment:
I would have stored the last_handled_blocked as well in here and did not get it from the db just for performance reasons and saving one query. But this is very unimportant, I guess.

Reply:
I don't think it makes a noticable difference and as you said I prefer this way because it's slightly cleaner.